source/adapters/hip/command_buffer.cpp

```diff
@@ -9,156 +9,756 @@
 //===----------------------------------------------------------------------===//
 
 #include "command_buffer.hpp"
+
+#include <hip/hip_runtime.h>
+
 #include "common.hpp"
+#include "enqueue.hpp"
+#include "event.hpp"
+#include "kernel.hpp"
+#include "memory.hpp"
+#include "queue.hpp"
+
+#include <cstring>
+
+ur_exp_command_buffer_handle_t_::ur_exp_command_buffer_handle_t_(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice)
+    : Context(hContext),
+      Device(hDevice), HIPGraph{nullptr}, HIPGraphExec{nullptr}, RefCount{1} {
+  urContextRetain(hContext);
+  urDeviceRetain(hDevice);
+}
+
+/// The ur_exp_command_buffer_handle_t_ destructor releases
+/// all the memory objects allocated for command_buffer managment
+ur_exp_command_buffer_handle_t_::~ur_exp_command_buffer_handle_t_() {
+  // Release the memory allocated to the Context stored in the command_buffer
+  UR_TRACE(urContextRelease(Context));
+
+  // Release the device
+  UR_TRACE(urDeviceRelease(Device));
+
+  // Release the memory allocated to the HIPGraph
+  UR_CHECK_ERROR(hipGraphDestroy(HIPGraph));
+
+  // Release the memory allocated to the HIPGraphExec
+  UR_CHECK_ERROR(hipGraphExecDestroy(HIPGraphExec));
+}
+
+/// Helper function for finding the HIP Nodes associated with the
+/// commands in a command-buffer, each event is pointed to by a sync-point in
+/// the wait list.
+///
+/// @param[in] CommandBuffer to lookup the events from.
+/// @param[in] NumSyncPointsInWaitList Length of \p SyncPointWaitList.
+/// @param[in] SyncPointWaitList List of sync points in \p CommandBuffer
+/// to find the events for.
+/// @param[out] HipNodesList Return parameter for the HIP Nodes associated with
+/// each sync-point in \p SyncPointWaitList.
+///
+/// @return UR_RESULT_SUCCESS or an error code on failure
+static ur_result_t getNodesFromSyncPoints(
+    const ur_exp_command_buffer_handle_t &CommandBuffer,
+    size_t NumSyncPointsInWaitList,
+    const ur_exp_command_buffer_sync_point_t *SyncPointWaitList,
```

Comment:
Is `SuncPointWaitList` passed as a pointer (as opposed to by reference )because it's an array? If so (and it's a general comment, not just for this func) would it be good to assert on it having a value set prior to using it. Or are we sure that the values will never be null?

Reply:
`SyncPointWaitList` is provided by the UR API and we forward it directly to this internal function without reformatting it.
We added asserts entry points functions to check the value of `pSyncPointWaitList`.